<h2>Projeto | Banco de Dados I</h2>

Aluno: Rafael Damasceno de Almeida | Id: 1116006

Curso: Data Science - Santander Coders 2023 | 2º Semestre

### Objetivo:

O objetivo deste projeto é aplicar o conteúdo ensinado na disciplina de **Banco de Dados 1**, aplicando em um Dataset comandos SQL para: construção do Database, carregamento de Dados, consulta e análise dos dados.

### Dataset:
Essa é uma base de Dados pública de pedidos feitos na loja "Olist". Ela possui informações sobre 100 mil pedidos feitos entre 2016 e 2018 em multiplos marketplaces brasileiros.     Seus elementos permitem visualizar um pedido em múltiplas dimensões, como: status do pedido, preço, pagamento, frete, localização do cliente, atributos do produto e até avaliações escritas pelos clientes.

Link: https://www.kaggle.com/code/iuyhnm/starter-brazilian-e-commerce-public-c805b9c6-d
### Esquema de Dados
Os dados são divididos entre multiplas tabelas, um esquema do relacionamento entre elas pode ser visto abaixo:

<img src="https://i.imgur.com/HRhd2Y0.png" width="600"/>



### Construção de Banco de Dados no PostgreSQ :

A montagem da base de dados foi feita no **PostgreSQL**. 

Após fazer o download dos arquivos do Dataset, que estavam em formato "CSV", e criação de uma nova Database nomeada "projeto", foi necessário criar uma tabela do PostgreSQL para cada uma das tabelas do Dataset, declarando o nome e o tipo de suas colunas. Utilizei o comando **COPY** para importar os dados dos arquivos CSV.

NOTA: Tentei utilizar a ferramenta da própria interface do pgAdmin4, porém, surgiram vários erros que dificultavam a importação.

Os comandos de criação das tabelas e importação dos dados usados no pgAdmin4 podem ser vistos abaixo:

In [ ]:
--Criando tabela customers
CREATE TABLE customers (
	customer_id varchar (50),
	customer_unique_id varchar(50),
	customer_zip_code int,
	customer_city varchar (50),
	customer_state varchar (50)
	)
COPY customers (customer_id, customer_unique_id, customer_zip_code, customer_city, customer_state)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_customers_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela geolocation
CREATE TABLE geolocation (
	geolocation_zip_code_prefix int,
	geolocation_lat float,
	geolocation_Ing float,
	geolocation_city varchar (50),
	geolocation_state varchar (10)
	)
COPY geolocation (geolocation_zip_code_prefix, geolocation_lat, geolocation_Ing, geolocation_city, geolocation_state)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_geolocation_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela order_items
CREATE TABLE order_items (
	order_id varchar,
	order_item_id varchar (50),
	product_id varchar (50),
	seller_id varchar (50),
	shipping_limit_date timestamp,
	price float,
	freight_value float
    ) 
COPY order_items (order_id, order_item_id, product_id, seller_id, shipping_limit_date, price,freight_value)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_order_items_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela order_payments
CREATE TABLE order_payments (
	order_id varchar(50), 
	payment_sequential int, 
	payment_type varchar (50), 
	payment_installments int, 
	payment_value float
    )
COPY order_payments (order_id, payment_sequential, payment_type, payment_installments, payment_value)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_order_payments_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela reviews
CREATE TABLE reviews (
    review_id varchar (50),
	order_id varchar (50), 
	review_score int, 
	review_comment_title varchar (50), 
	review_comment_message varchar (500), 
	review_creation_date timestamp, 
	review_answer_timestamp timestamp
    )
COPY reviews (review_id, order_id, review_score, review_comment_title, review_comment_message, review_creation_date, review_answer_timestamp)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_order_reviews_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela orders
CREATE TABLE orders (
	order_id VARCHAR(100),
	customer_id VARCHAR(100),
	order_status VARCHAR(50),
	order_purchase_timestamp TIMESTAMP,
	order_approved_at TIMESTAMP,
	order_delivered_carrier_date TIMESTAMP,
	order_delivered_customer_date TIMESTAMP,
	order_estimated_delivery_date TIMESTAMP
	) 
COPY orders (order_id, customer_id, order_status, order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date,order_estimated_delivery_date)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_orders_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela products
CREATE TABLE products (
	product_id varchar(50),
	product_category_name VARCHAR(70),
	product_name_length int ,
	product_description_length int,
	product_photos_qty int,
	product_weight_g int ,
	product_length_cm int ,
	product_height_cm int ,
	product_width_cm int 
    ) 
COPY products (product_id, product_category_name, product_name_length, product_description_length, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_products_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela sellers
CREATE TABLE sellers(
    seller_id varchar(50),
	seller_zip_code_prefix int,
	seller_city varchar (100) ,
	seller_state varchar(15)
	)
COPY sellers (seller_id, seller_zip_code_prefix, seller_city, seller_state)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\olist_sellers_dataset.csv' DELIMITER ',' CSV HEADER;

In [ ]:
--Criando tabela product_category_name_translation
CREATE TABLE product_category_name_translation (
	product_category_name varchar(50),
	product_category_name_english varchar(70)
	)
COPY product_category_name_translation (product_category_name, product_category_name_english)
FROM 'G:\Aulas santander\banco de dados\projeto banco de dados\Dataset\product_category_name_translation.csv' DELIMITER ',' CSV HEADER;

### Acessando o Banco de Dados criado através de bibliotecas Python
Utilizaremos a função "create_engine" da biblioteca "sqlalchemy", que cria um objeto de conexão com um banco de dados relacional. Este objeto de conexão é usado para se comunicar com o banco de dados e executar consultas SQL. Ele segue o padrão  "SGBD://USER:SENHA@HOST/DATABASE" que deve ser preenchido com as informações do seu banco.

In [2]:
from sqlalchemy import create_engine
import pandas as pd
SGBD = "postgresql+psycopg2"
USER = "postgres"
SENHA = "admin"
HOST = "localhost"
DATABASE = "projeto"

engine = create_engine(f"{SGBD}://{USER}:{SENHA}@{HOST}:5432/{DATABASE}")
engine.connect()

### Exibição e descrição dos dados
Usarei o comando *read_sql_query* da biblioteca Pandas para ler os dados. Isso permite que você execute consultas SQL diretamente em um banco de dados e carregue os resultados em um DataFrame do Pandas, oferecendo uma maneira eficiente de carregar dados.

Com os dados carregados em um DataFrame Pandas é possivel usar o método .info(), que fornece um resumo conciso das informações sobre o DataFrame, incluindo:

- O índice: Tipo de índice, quantidade de entradas não nulas e tamanho do índice.
- As colunas: Nome das colunas, quantidade de entradas não nulas em cada coluna e tipo de dados de cada coluna.
- A quantidade total de memória usada pelo DataFrame.

Abaixo exibirei separadamente todas as tabelas, seguidas de suas descrições com o método .info:

In [7]:
query = """
SELECT *
FROM
public.customers
"""
pd.read_sql_query(query, con = engine)


customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code          customer_city customer_state  
0                  14409                 franca             SP  
1                   9790  sao bernardo do campo             SP  
2                   1151              sao paulo             SP  
3                   8775        mogi das cruzes             SP  
4                  13056               campinas             SP  
...                  ...                    ...            ...  
99436               3937              sao paulo             SP  
99437               6764        taboao da serra             SP  
99438              60115              fortaleza             CE  
99439              92120                 canoas             RS  
99440               6703                  cotia             SP  

[99441 rows x 5 columns]

Tabela customers: Essa tabela apresenta os dados cadastrais dos clientes, contendo: ID, prefixo do código postal, cidade e estado.

In [110]:
query = """
SELECT *
FROM
public.customers
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         99441 non-null  object
 1   customer_unique_id  99441 non-null  object
 2   customer_zip_code   99441 non-null  int64 
 3   customer_city       99441 non-null  object
 4   customer_state      99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None


In [8]:
query = """
SELECT *
FROM
public.geolocation
"""
pd.read_sql_query(query, con = engine)

geolocation_zip_code_prefix  geolocation_lat  geolocation_ing  \
0                               1037       -23.545621       -46.639292   
1                               1046       -23.546081       -46.644820   
2                               1046       -23.546129       -46.642951   
3                               1041       -23.544392       -46.639499   
4                               1035       -23.541578       -46.641607   
...                              ...              ...              ...   
1000158                        99950       -28.068639       -52.010705   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000161                        99980       -28.388932       -51.846871   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS  

[1000163 rows x 5 columns]

Tabela geolocation: Essa tabela apresenta as informações de cada prefixo postal, contendo: latitude, longitude, cidade e estado.

In [111]:
query = """
SELECT *
FROM
public.geolocation
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_ing              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB
None


In [5]:
query = """
SELECT *
FROM
public.order_items
ORDER BY order_id
"""
pd.read_sql_query(query, con = engine)

order_id order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214             1   
1       00018f77f2f0320c557190d7a144bdd3             1   
2       000229ec398224ef6ca0657da4fc703e             1   
3       00024acbcdf0a6daa1e931b038114c75             1   
4       00042b26cf59d7ce69dfabb4e55b4fd9             1   
...                                  ...           ...   
112645  fffc94f6ce00a00581880bf54a75a037             1   
112646  fffcd46ef2263f404302a634eb57f7eb             1   
112647  fffce4705a9662cd70adb13d4a31832d             1   
112648  fffe18544ffabc95dfada21779c9644f             1   
112649  fffe41c64501cc87c801fd61db3f6244             1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

       shipping_limit_date   price  freight_value  
0      2017-09-19 09:45:35   58.90          13.29  
1      2017-05-03 11:05:13  239.90          19.93  
2      2018-01-18 14:48:30  199.00          17.87  
3      2018-08-15 10:10:18   12.99          12.79  
4      2017-02-13 13:57:51  199.90          18.14  
...                    ...     ...            ...  
112645 2018-05-02 04:11:01  299.99          43.41  
112646 2018-07-20 04:31:48  350.00          36.53  
112647 2017-10-30 17:14:25   99.90          16.95  
112648 2017-08-21 00:04:32   55.99           8.72  
112649 2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

Tabela order_items: Essa tabela apresenta os dados de todos os produtos vendidos, separadamente, contendo: ID do pedido, ID do produto, ID do vendedor, data limite de entrega, preço do produto e do frete.

In [112]:
query = """
SELECT *
FROM
public.order_items
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  object        
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 6.0+ MB
None


In [8]:
query = """
SELECT *
FROM
public.order_payments
ORDER BY order_id
"""
pd.read_sql_query(query, con = engine)

order_id  payment_sequential payment_type  \
0       00010242fe8c5a6d1ba2dd792cb16214                   1  credit_card   
1       00018f77f2f0320c557190d7a144bdd3                   1  credit_card   
2       000229ec398224ef6ca0657da4fc703e                   1  credit_card   
3       00024acbcdf0a6daa1e931b038114c75                   1  credit_card   
4       00042b26cf59d7ce69dfabb4e55b4fd9                   1  credit_card   
...                                  ...                 ...          ...   
103881  fffc94f6ce00a00581880bf54a75a037                   1       boleto   
103882  fffcd46ef2263f404302a634eb57f7eb                   1       boleto   
103883  fffce4705a9662cd70adb13d4a31832d                   1  credit_card   
103884  fffe18544ffabc95dfada21779c9644f                   1  credit_card   
103885  fffe41c64501cc87c801fd61db3f6244                   1  credit_card   

        payment_installments  payment_value  
0                          2          72.19  
1                          3         259.83  
2                          5         216.87  
3                          2          25.78  
4                          3         218.04  
...                      ...            ...  
103881                     1         343.40  
103882                     1         386.53  
103883                     3         116.85  
103884                     3          64.71  
103885                     1          55.79  

[103886 rows x 5 columns]

Tabela order_payments: Essa tabela apresenta os dados de todos os pedidos feitos, contendo: ID do pedido, tipo de pagamento e valor do pagamento.

In [113]:
query = """
SELECT *
FROM
public.order_payments
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None


In [11]:
query = """
SELECT *
FROM
public.orders
"""
pd.read_sql_query(query, con = engine)

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0                        2017-10-18  
1                        2018-08-13  
2                        2018-09-04  
3                        2017-12-15  
4                        2018-02-26  
...                             ...  
99436                    2017-03-28  
99437                    2018-03-02  
99438                    2017-09-27  
99439                    2018-02-15  
99440                    2018-04-03  

[99441 rows x 8 columns]

Tabela orders: Essa tabela apresenta os dados sobre o estado da entrega dos pedidos, contendo: ID do pedido, ID do cliente, estado da entrega, data de postagem, data estimada da entrega e data da entrega.

In [114]:
query = """
SELECT *
FROM
public.orders
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB
None


In [13]:
query = """
SELECT *
FROM
public.product_category_name_translation
"""
pd.read_sql_query(query, con = engine)

product_category_name product_category_name_english
0                    beleza_saude                 health_beauty
1          informatica_acessorios         computers_accessories
2                      automotivo                          auto
3                 cama_mesa_banho                bed_bath_table
4                moveis_decoracao               furniture_decor
..                            ...                           ...
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

[71 rows x 2 columns]

Tabela product_category_name_translation: Essa tabela traduz o nome do produto em português para o inglês.

In [115]:
query = """
SELECT *
FROM
public.product_category_name_translation
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB
None


In [14]:
query = """
SELECT *
FROM
public.products
"""
pd.read_sql_query(query, con = engine)

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_length  product_description_length  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

Tabela products: Essa tabela apresenta os dados de descriçào de todos os produtos, contendo: ID do produto, categoria do produto e informações de tamanho e peso.

In [116]:
query = """
SELECT *
FROM
public.products
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_length         32341 non-null  float64
 3   product_description_length  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None


In [15]:
query = """
SELECT *
FROM
public.reviews
"""
pd.read_sql_query(query, con = engine)

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                 None   
1                 5                 None   
2                 5                 None   
3                 5                 None   
4                 5                 None   
...             ...                  ...   
99219             5                 None   
99220             5                 None   
99221             5                 None   
99222             4                 None   
99223             1                 None   

                                  review_comment_message review_creation_date  \
0                                                   None           2018-01-18   
1                                                   None           2018-03-10   
2                                                   None           2018-02-17   
3                  Recebi bem antes do prazo estipulado.           2017-04-21   
4      Parabéns lojas lannister adorei comprar pela I...           2018-03-01   
...                                                  ...                  ...   
99219                                               None           2018-07-07   
99220                                               None           2017-12-09   
99221  Excelente mochila, entrega super rápida. Super...           2018-03-22   
99222                                               None           2018-07-01   
99223  meu produto chegou e ja tenho que devolver, po...           2017-07-03   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

Tabela reviews: Essa tabela apresenta os dados de todos reviews feitos sobre as compras, contendo: nota, a mensagem comentada, data de criação do review e data que foi respondido.

In [117]:
query = """
SELECT *
FROM
public.reviews
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  object        
 4   review_comment_message   40977 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB
None


In [16]:
query = """
SELECT *
FROM
public.sellers
"""
pd.read_sql_query(query, con = engine)

seller_id  seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                   13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4     51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...                                ...                     ...   
3090  98dddbc4601dd4443ca174359b237166                   87111   
3091  f8201cab383e484733266d1906e2fdfa                   88137   
3092  74871d19219c7d518d0090283e03c137                    4650   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

Tabela sellers: Essa tabela apresenta os dados de todos vendedores e de onde são, contendo: prefixo postal, cidade e estado do vendedor.

In [118]:
query = """
SELECT *
FROM
public.sellers
"""
dataframe = pd.read_sql_query(query, con = engine)
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB
None


### Análise exploratória

Verificando quais categorias possuem mais produtos:

In [93]:
query = """
SELECT 
    product_category_name AS Categoria,
    COUNT(product_category_name) AS quantidade
FROM
public.products
GROUP BY product_category_name
ORDER BY quantidade DESC
"""
pd.read_sql_query(query, con = engine)

categoria  quantidade
0         cama_mesa_banho        3029
1           esporte_lazer        2867
2        moveis_decoracao        2657
3            beleza_saude        2444
4   utilidades_domesticas        2335
..                    ...         ...
69        casa_conforto_2           5
70               pc_gamer           3
71     seguros_e_servicos           2
72      cds_dvds_musicais           1
73                   None           0

[74 rows x 2 columns]

A categoria que possui mais produtos é "cama_mesa_banho", com 3029 produtos.


---
Verificando ID do produto mais vendido:

In [24]:
query = """
SELECT 
    product_id AS id_produto,
    count(product_id) AS numero_vendas

FROM
public.order_items
GROUP BY product_id
ORDER BY numero_vendas DESC
"""

pd.read_sql_query(query, con = engine)

id_produto  numero_vendas
0      aca2eb7d00ea1a7b8ebd4e68314663af            527
1      99a4788cb24856965c36a24e339b6058            488
2      422879e10f46682990de24d770e7f83d            484
3      389d119b48cf3043d311335e499d9c6b            392
4      368c6c730842d78016ad823897a372db            388
...                                 ...            ...
32946  9eb0c62c8c4159426b700eb10045dc44              1
32947  39c4b45f85d57630198b4da1688c13a0              1
32948  580af1ebaa84e22ae14e08f073d091b5              1
32949  cf30110b1e85017c00752838fec35442              1
32950  2a0d1fb3e64019c7cf0e75d1cf336277              1

[32951 rows x 2 columns]

O ID do produto mais vendido é "aca2eb7d00ea1a7b8ebd4e68314663af", com 527 vendas.

---
Verificando estado e cidade com mais clientes únicos:

In [159]:
#Estado com mais clientes únicos
query = """
SELECT
    customer_state AS estado_cliente,
    COUNT(DISTINCT customer_unique_id) AS clientes
FROM
public.customers
GROUP BY customer_state
ORDER BY clientes DESC
"""

pd.read_sql_query(query, con=engine)

estado_cliente  clientes
0              SP     40302
1              RJ     12384
2              MG     11259
3              RS      5277
4              PR      4882
5              SC      3534
6              BA      3277
7              DF      2075
8              ES      1964
9              GO      1952
10             PE      1609
11             CE      1313
12             PA       949
13             MT       876
14             MA       726
15             MS       694
16             PB       519
17             PI       482
18             RN       474
19             AL       401
20             SE       342
21             TO       273
22             RO       240
23             AM       143
24             AC        77
25             AP        67
26             RR        45

O estado com mais clientes únicos é SP, com 40302 clientes.

In [157]:
#Cidade com mais clientes únicos
query = """
SELECT 
    customer_city AS cidade_cliente,
    customer_state AS estado_cliente,
    COUNT(DISTINCT customer_unique_id) AS clientes
FROM
public.customers
GROUP BY customer_city, customer_state
ORDER BY clientes DESC
"""
pd.read_sql_query(query, con = engine)


cidade_cliente estado_cliente  clientes
0                  sao paulo             SP     14984
1             rio de janeiro             RJ      6620
2             belo horizonte             MG      2672
3                   brasilia             DF      2069
4                   curitiba             PR      1465
...                      ...            ...       ...
4305           santa brigida             BA         1
4306           santa cecilia             PB         1
4307  santa cecilia do pavao             PR         1
4308     santa clara d'oeste             SP         1
4309          candido mendes             MA         1

[4310 rows x 3 columns]

A cidade com mais clientes únicos é São Paulo, com 14984 clientes únicos.

---
Relacionando tabela "sellers" à tabela "order_items", para saber quais são e de onde são os vendedores que mais arrecadaram:

In [25]:
query = """
SELECT
    s.seller_id,
    seller_state,
    seller_city,
    SUM(o.price) AS total_vendido
FROM
public.sellers AS s
LEFT JOIN public.order_items AS o
ON s.seller_id = o.seller_id
GROUP BY s.seller_id
ORDER BY total_vendido DESC
"""
pd.read_sql_query(query, con = engine)

seller_id seller_state       seller_city  \
0     4869f7a5dfa277a7dca6462dcf3b52b2           SP           guariba   
1     53243585a1d6dc2643021fd1853d8905           BA  lauro de freitas   
2     4a3ca9315b744ce9f8e9374361493884           SP          ibitinga   
3     fa1c13f2614d7b5c4749cbc52fecda94           SP            sumare   
4     7c67e1448b00f6e969d365cea6b010ab           SP   itaquaquecetuba   
...                                ...          ...               ...   
3090  34aefe746cd81b7f3b23253ea28bef39           PR          curitiba   
3091  702835e4b785b67a084280efca355756           MG      juiz de fora   
3092  1fa2d3def6adfa70e58c276bb64fe5bb           SP         sao paulo   
3093  77128dec4bec4878c37ab7d6169d6f26           SP         sao paulo   
3094  cf6f6bc4df3999b9c6440f124fb2f687           SP         sao paulo   

      total_vendido  
0         229472.63  
1         222776.05  
2         200472.92  
3         194042.03  
4         187923.89  
...             ...  
3090           8.00  
3091           7.60  
3092           6.90  
3093           6.50  
3094           3.50  

[3095 rows x 4 columns]

O id do vendedor que mais arrecadou foi "4869f7a5dfa277a7dca6462dcf3b52b2	", da cidade de Guariba-SP, que vendeu um total de R$229.472,63.


---
Relacionando tabelas "products" e "order_items", para saber quais categorias de produtos que mais arrecadaram:

In [26]:

query= '''
SELECT 
    p.product_category_name AS categoria_produto,
    SUM(o.price) AS valor_vendido

FROM public.products AS p
LEFT JOIN public.order_items AS o
ON p.product_id = o.product_id
GROUP BY p.product_category_name
ORDER BY valor_vendido DESC
'''
pd.read_sql_query(query, con = engine)

categoria_produto  valor_vendido
0                    beleza_saude     1258681.34
1              relogios_presentes     1205005.68
2                 cama_mesa_banho     1036988.68
3                   esporte_lazer      988048.97
4          informatica_acessorios      911954.32
..                            ...            ...
69                         flores        1110.04
70                casa_conforto_2         760.27
71              cds_dvds_musicais         730.00
72  fashion_roupa_infanto_juvenil         569.85
73             seguros_e_servicos         283.29

[74 rows x 2 columns]

A categoria que mais arrecadou foi "beleza_saude", com o valor vendido de R$1.258.681,34.

---
Relacionando as tabelas "orders" e "customers", para saber a data do primeiro pedido de cada cliente:

In [27]:

query='''
SELECT c.customer_unique_id AS id_cliente, 
	MIN(o.order_purchase_timestamp) AS data_primeiro_pedido
FROM public.orders AS o 
LEFT JOIN public.customers AS c 
ON c.customer_id = o.customer_id
GROUP BY c.customer_unique_id
ORDER BY c.customer_unique_id DESC
'''
pd.read_sql_query(query, con = engine)

id_cliente data_primeiro_pedido
0      ffffd2657e2aad2907e67c3e9daecbeb  2017-05-02 20:18:45
1      ffff5962728ec6157033ef9805bacc48  2018-05-02 15:17:41
2      ffff371b4d645b6ecea244b27531430a  2017-02-07 15:49:16
3      fffea47cd6d3cc0a88bd621562a9d061  2017-12-10 20:07:56
4      fffcf5a5ff07b0908bd4e2dbc735a684  2017-06-08 21:00:36
...                                 ...                  ...
96091  0004aac84e0df4da2b147fca70cf8255  2017-11-14 19:45:42
96092  0000f6ccb0745a6a4b88665a16c9f078  2017-10-12 20:29:41
96093  0000f46a3911fa3c0805444483337064  2017-03-10 21:05:03
96094  0000b849f77a49e4a4ce2b2a4ca5be3f  2018-05-07 11:11:27
96095  0000366f3b9a7992bf8c76cfdf3221e2  2018-05-10 10:56:27

[96096 rows x 2 columns]

---
Relacionando tabelas "customers", "orders" e "order_payments", para saber qual foi o cliente que mais gastou e de onde ele é:

In [10]:

query ='''
SELECT
    c.customer_unique_id AS id_cliente,
    c.customer_city AS cidade,
    c.customer_state AS estado,
    SUM(op.payment_value) AS total_pago
FROM public.orders AS o
LEFT JOIN public.customers AS c
ON c.customer_id = o.customer_id
LEFT JOIN public.order_payments AS op
ON op.order_id = o.order_id
GROUP BY c.customer_unique_id, c.customer_city, c.customer_state
HAVING NOT SUM(op.payment_value) IS NULL
ORDER BY total_pago DESC
'''
pd.read_sql_query(query, con = engine)

id_cliente          cidade estado  total_pago
0      0a0a92112bd4c708ca5fde585afaa872  rio de janeiro     RJ    13664.08
1      46450c74a0d8c5ca9395da1daac6c120   florianopolis     SC     9553.02
2      da122df9eeddfedc1dc1f5349a1a690c        araruama     RJ     7571.63
3      763c8b1c9c68a0229c42c9fc6f662b93      vila velha     ES     7274.88
4      dc4802a71eae9be1dd28f5d788ceb526    campo grande     MS     6929.31
...                                 ...             ...    ...         ...
96213  b33336f46234b24a613ad9064d13106d       sao paulo     SP       10.89
96214  bd06ce0e06ad77a7f681f1a4960a3cc6       sao paulo     SP       10.07
96215  317cfc692e3f86c45c95697c61c853a6        paulinia     SP        9.59
96216  4fa4365000c7090fcb8cad5713c6d3db       sao paulo     SP        0.00
96217  968fac81e2c44fb6c1e3ac2a45e6a102       sao paulo     SP        0.00

[96218 rows x 4 columns]

Usando Subquery para descobrir valor médio pago por cliente:

In [29]:
#Valor médio pago por clientes
query ='''
WITH gastos_clientes AS(
    SELECT
    c.customer_unique_id,
    c.customer_city,
    c.customer_state,
    SUM(op.payment_value) AS total_pago
FROM public.orders AS o
LEFT JOIN public.customers AS c
ON c.customer_id = o.customer_id
LEFT JOIN public.order_payments AS op
ON op.order_id = o.order_id
GROUP BY c.customer_unique_id, c.customer_city, c.customer_state
ORDER BY total_pago DESC
)
SELECT CAST(AVG(total_pago) AS NUMERIC(10, 2)) AS media_total_pago
FROM gastos_clientes

'''
pd.read_sql_query(query, con = engine)

media_total_pago
0            166.38

O Valor médio pago por cliente foi de R$166,38.

---
Vendedores que possuem a nota média de suas reviews iguais a 5:

In [63]:
query ='''
	SELECT 
    o.seller_id AS id_vendedor,
    AVG(review_score) AS nota_media
FROM public.reviews AS r
LEFT JOIN order_items AS o
ON o.order_id = r.order_id
GROUP BY o.seller_id
HAVING AVG(review_score) = 5
ORDER BY nota_media DESC

'''
pd.read_sql_query(query, con = engine)

id_vendedor  nota_media
0    a61cc04793308395a840807104365121         5.0
1    be67f78487e2cecb0d55bc769709e4f5         5.0
2    c8143b3069f6746a77421b5ce30a450c         5.0
3    2a73cba571d90c694b7caca072ccf6ce         5.0
4    6c17baf138731a4aaacc3210adf2037b         5.0
..                                ...         ...
502  16bdc8cefd0e32a6f0824d296c5ad14a         5.0
503  48b6c3f4c6a93171da04b75313f2130f         5.0
504  b2eecf5ea250510da76590ca79d60e5d         5.0
505  eae9af4811c294c56795d70e715b7337         5.0
506  1c83418358f249c9d6affa52d579db0e         5.0

[507 rows x 2 columns]

### Conclusão

A realização deste projeto me proporcionou uma valiosa oportunidade para consolidar os conhecimentos adquiridos no módulo "Banco de Dados I" do curso de Data Science. Através da importação de uma base de dados CSV para uma base de dados PostgreSQL, pude não apenas praticar as técnicas de manipulação e análise de dados, mas também explorar diversas consultas fundamentais, e através da análise exploratória respondi diferentes questões, que contribuíram significativamente para a minha compreensão prática dos conceitos apresentados no curso.